dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp

```diff
@@ -390,44 +446,63 @@ FastAddPeerRes FastAddPeerImpl(
 {
     auto log = Logger::get("FastAddPeer");
-    LOG_INFO(log, "Begin apply fap snapshot, region_id={}, peer_id={}", region_id, peer_id);
-    GET_METRIC(tiflash_fap_task_state, type_ingesting_stage).Increment();
-    SCOPE_EXIT({ GET_METRIC(tiflash_fap_task_state, type_ingesting_stage).Decrement(); });
     Stopwatch watch_ingest;
     auto kvstore = tmt.getKVStore();
     auto fap_ctx = tmt.getContext().getSharedContextDisagg()->fap_context;
     auto checkpoint_ingest_info = fap_ctx->getOrRestoreCheckpointIngestInfo(tmt, proxy_helper, region_id, peer_id);
-    kvstore->handleIngestCheckpoint(checkpoint_ingest_info->getRegion(), checkpoint_ingest_info, tmt);
-    fap_ctx->cleanCheckpointIngestInfo(tmt, region_id);
-    GET_METRIC(tiflash_fap_task_duration_seconds, type_ingest_stage).Observe(watch_ingest.elapsedSeconds());
-    auto begin = checkpoint_ingest_info->beginTime();
-    auto current = FAPAsyncTasks::getCurrentMillis();
-    if (begin != 0)
+    if (!checkpoint_ingest_info)
     {
-        GET_METRIC(tiflash_fap_task_duration_seconds, type_total).Observe((current - begin) / 1000.0);
+        // If fap is enabled, and this region is not currently exists on proxy's side,
+        // proxy will check if we have a fap snapshot first.
+        // If we don't, the snapshot should be a regular snapshot.
+        LOG_DEBUG(
+            log,
+            "Failed to get fap snapshot, it's regular snapshot, region_id={}, peer_id={}",
+            region_id,
+            peer_id);
+        return false;
+    }
+    LOG_INFO(log, "Begin apply fap snapshot, region_id={}, peer_id={}", region_id, peer_id);
+    GET_METRIC(tiflash_fap_task_result, type_restore).Increment();
+    // If there is `checkpoint_ingest_info`, it is exactly the data we want to ingest. Consider two scene:
+    // 1. If there was a failed FAP which failed to clean, its data will be overwritten by current FAP which has finished phase 1.
+    // 2. It is not possible that a restart happens at FAP phase 2, and a regular snapshot is sent, because snapshots can only be accepted once the previous snapshot it handled.
+    {
+        GET_METRIC(tiflash_fap_task_state, type_ingesting_stage).Increment();
+        SCOPE_EXIT({ GET_METRIC(tiflash_fap_task_state, type_ingesting_stage).Decrement(); });
+        kvstore->handleIngestCheckpoint(checkpoint_ingest_info->getRegion(), checkpoint_ingest_info, tmt);
+        fap_ctx->cleanTask(tmt, proxy_helper, region_id, true);
+        GET_METRIC(tiflash_fap_task_duration_seconds, type_ingest_stage).Observe(watch_ingest.elapsedSeconds());
+        auto begin = checkpoint_ingest_info->beginTime();
+        auto current = FAPAsyncTasks::getCurrentMillis();
+        if (begin != 0)
+        {
+            GET_METRIC(tiflash_fap_task_duration_seconds, type_total).Observe((current - begin) / 1000.0);
+        }
+        LOG_INFO(
+            log,
+            "Finish apply fap snapshot, region_id={} peer_id={} begin_time={} current_time={}",
```

Comment:
log down the time cost is better than log down the begin_time and current_time

Reply:
when begin_time != 0, log down the time cost; when begin_time == 0, log down a "unknown"?

Reply:
The begin time here is to find where the FAP snapshot is generated

Reply:
I would like to keep the begin_time in the log of `Begin apply fap snapshot`